src/date-range-picker/calendar/grids/grid.tsx

```diff
@@ -0,0 +1,335 @@
+ * - (keyboard navigation) Safari/Chrome+VO - days are not announced as interactive (clickable or selectable);
+ * - (keyboard navigation) Safari/Chrome+VO - day announcements are not interruptive and can be missed if navigating fast.
+ */
+
+export function Grid({
+  baseDate,
+  selectedStartDate,
+  selectedEndDate,
+  rangeStartDate,
+  rangeEndDate,
+  focusedDate,
+
+  focusedDateRef,
+
+  onSelectDate,
+  onGridKeyDownHandler,
+  onFocusedDateChange,
+
+  isDateEnabled,
+  dateDisabledReason,
+
+  locale,
+  todayAriaLabel,
+  currentMonthAriaLabel,
+  ariaLabelledby,
+  className,
+  startOfWeek = 0,
+  granularity = 'day',
+}: GridProps) {
+  const baseDateTime = baseDate?.getTime();
+  // `baseDateTime` is used as a more stable replacement for baseDate
+  const weeks = useMemo<Date[][]>(
+    () => getCalendarMonth(baseDate, { firstDayOfWeek: startOfWeek }),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [baseDateTime, startOfWeek]
+  );
+  const i18n = useInternalI18n('date-range-picker');
+  const isMonthPicker = granularity === 'month';
+  const weekdays = weeks[0].map(date => date.getDay());
+  const normalizedLocale = normalizeLocale('DateRangePicker', locale ?? null);
+  const quarters = useCalendarGridRows({ baseDate, granularity: 'month', locale: normalizedLocale, startOfWeek });
+  const rows = isMonthPicker ? quarters : weeks;
+
+  const currentAnnouncement = i18n(
+    granularity === 'day' ? 'i18nStrings.todayAriaLabel' : 'i18nStrings.currentMonthAriaLabel',
+    granularity === 'day' ? todayAriaLabel : currentMonthAriaLabel
+  );
+
+  return (
+    <table role="grid" aria-labelledby={ariaLabelledby} className={clsx(styles.grid, className)}>
+      {!isMonthPicker && (
+        <thead>
+          <tr>
+            {weekdays.map(dayIndex => (
+              <th key={dayIndex} scope="col" className={clsx(styles['grid-cell'], testutilStyles['day-header'])}>
+                <span aria-hidden="true">{renderDayName(locale, dayIndex, 'short')}</span>
+                <ScreenreaderOnly>{renderDayName(locale, dayIndex, 'long')}</ScreenreaderOnly>
+              </th>
+            ))}
+          </tr>
+        </thead>
+      )}
+      <tbody onKeyDown={onGridKeyDownHandler}>
+        {rows.map((row, rowIndex) => {
+          return (
+            <tr key={rowIndex} className={isMonthPicker ? testutilStyles.quarter : testutilStyles.week}>
+              {row.map((date, rowItemIndex) => {
+                const itemKey = isMonthPicker
+                  ? `Month ${rowIndex * 3 + rowItemIndex + 1}`
+                  : `${rowIndex}:${rowItemIndex}`;
+                const isSameItem = isMonthPicker ? isSameMonth : isSameDay;
+                const isSamePage = isMonthPicker ? isSameYear : isSameMonth;
+                const addItems = isMonthPicker ? addMonths : addDays;
+                const addRows = isMonthPicker ? addQuarters : addWeeks;
+                const pageName = isMonthPicker ? 'year' : 'month';
```

Comment:
Time to call for strategy pattern: https://en.wikipedia.org/wiki/Strategy_pattern

```js
const { isSameItem, isSamePage, .... } = isMonthPicker ? monthNavigationHelpers : dayNavigationHelpers
```

Reply:
After reading to the end, found more places to benefit from the strategy pattern. 

Up to you how many of them to fix, but IMO, all more than 3 repeating `isMonthPicker` or `granularity` condition in one function/scope can be replaced with this pattern from above